c++/itertools/itertools.hpp

```diff
@@ -61,24 +60,30 @@ namespace itertools {
       return c;
     }
 
-#if __cplusplus > 201703L
-    bool operator==(iterator_facade const &other) const { return self().equal(other.self()); }
-
-#else
-    template <typename U>
-    bool operator==(U const &other) const {
-      return self().equal(other);
-    }
+    decltype(auto) operator*() const { return self().dereference(); }
+    decltype(auto) operator->() const { return operator*(); }
+  };
 
-    template <typename U>
-    bool operator!=(U const &other) const {
-      return (!operator==(other));
+  template <typename Iter, typename EndIter>
+  inline typename std::iterator_traits<Iter>::difference_type distance(Iter first, EndIter last) {
+    if constexpr (std::is_same_v<typename std::iterator_traits<Iter>::iterator_category, std::random_access_iterator_tag>) {
```

Comment:
Ok, as we said, there will be a bug here if the original iterator is 
random access...
so for a simple `prod < vector<int>, vector<int> >`, it will fail ? test ?

Reply:
It will not fail for `std::vector` as here the type of `std::end` is the same as for `std::begin`.
We will need a proper random access range with a sentinel_t to test this.